File: cpp/data_structures/eight_chapter/Splay.c
#include "Splay.h"

template <typename NodePosi> inline
void attachAsLChild(NodePosi p, NodePosi lc) {
	p->lc = lc;
	if (lc)
		lc->parent = p;
}

template <typename NodePosi> inline
void attachAsRChild(NodePosi p, NodePosi rc) {
	p->rc = rc;
	if (rc)
		rc->parent = p;
}

template <typename T>
BinNodePosi(T) Splay<T>::splay(BinNodePosi(T) v) {
	if (!v)
		return nullptr;	
	BinNodePosi(T) p, g;
	while ((p = v->parent) && (g = p->parent)) {
		BinNodePosi(T) gg = g->parent;
		if (IsLChild(*v))
			if (IsLChild(*p)) {
				attachAsLChild(g, p->rc);
				attachAsLChild(p, v->rc);
				attachAsRChild(p, g);
				attachAsRChild(v, p);
			} else {
				attachAsRChild(g, v->lc);
				attachAsLChild(p, v->rc);
				attachAsLChild(v, g);
				attachAsRChild(v, p);
			}
		else if (IsRChild(*p)) {
			attachAsRChild(g, p->lc);
			attachAsLChild(p, g);
			attachAsRChild(p, v->lc);
			attachAsLChild(v, g);
		} else {
			attachAsRChild(v, g);
			attachAsLChild(v, p);
			attachAsRChild(p, v->lc);
			attachAsLChild(g, v->rc);
		}
		if (!gg)
			v->parent = nullptr;
		else
			(g == gg->lc) ? attachAsLRchild(gg, v) : attachAsRChild(gg, v);
		updateHeightaAbove(g);
	}
	if (p = v->parent) {
		if (IsLChild(*v)) {
			attachAsRChild(v, p);
			attachAsLChild(p, v->rc);
		} else {
			attachAsLChild(v, p);
			attachAsRChild(g, v->lc);
		}
		updateHeight(p);
		updateHeight(v);
	}
	v->parent = nullptr;
	return v;
}
			
template <typename T>
BinNodePosi(T) &Splay<T>::search(const T& e) {
	BinNodePosi(T) p = searchIn(this->_root, e, this->_hot = nullptr);
	this->_root = splay(p ? p : this->_hot);
	return this->_root;
}

template <typename T>
BinNodePosi(T) Splay<T>::insert(const T& e) {
	if (!this->_root) {
		this->_size++;
		return this->_root = new BinNode<T>(e);
	}
	if (e == search(e)->data)
		return this->_root;
	this->_size++;
	BinNodePosi(T) t = this->_root;   
	if (this->_root->data < e) {
		t->parent = this->_root = new BinNode<T>(e, nullptr, t, t->rc);
		if (HasRChild(*t)) {
			t->rc->parent = this->_root;
			t->rc = nullptr;
		}
	} else {
			t->parent = this->_root = new BinNode<T>(e, nullptr, t->lc, t);
			if (HasLChild(*t)) {
				t->lc->parent = this->_root;
				t->lc = nullptr;
			}
	}
	updateHeightAbove(t);
	return this->_root;
}

template <typename T>
bool Splay<T>::remove(const T& e) {
	if (!this->_root || (e != search(e)->data))
		return false;
	BinNodePosi(T) w = this->_root;
	if (!HasLChild(*this->_root)) {
		this->_root = this->_root->rc;
		if (this->_root)
			this->_root->parent = nullptr;
	} else if (!HasRChld(*this->_root)) {
		this->_root = this->_root->lc;
		if (this->_root)
			this->_root->parent = nullptr;
	} else {
		BinNodePosi(T) lTree = this->_root->lc;
		lTree->parent = nullptr;
		this->_root->lc = nullptr;
		search(w->data);
		this->_root->lc = lTree;
		lTree->parent = this->_root;
	}
	release(w->data);
	release(w);
	this->_size--;
	if (this->_root)
		updateHeight(this->_root);
	return true;
}

File: go/data_structure/6_chapter/splay/splay.go
package splay

import (
	"data_structure/5_chapter/binTree"
	"data_structure/5_chapter/bst"
)

type Splay struct {
	bst.BST
}

func (s *Splay) splay(n *binTree.BinNode) *binTree.BinNode {
	if n == nil {
		return nil
	}
	p := n.Parent
	if p == nil {
		return n
	}
	g := p.Parent
	for n.Parent != nil && n.Parent.Parent != nil {
		p = n.Parent
		g = p.Parent
		gg := g.Parent // great-grand parent
		if p.IsLChild() {
			if n.IsLChild() { // zig-zig
				attachAsLChild(g, p.Rc)
				attachAsLChild(p, n.Rc)
				attachAsRChild(p, g)
				attachAsRChild(n, p)
			} else { // zag-zig
				attachAsRChild(p, n.Lc)
				attachAsLChild(g, n.Rc)
				attachAsLChild(n, p)
				attachAsRChild(n, g)
			}
		} else {
			if n.IsLChild() { // zig-zag
				attachAsRChild(g, n.Lc)
				attachAsLChild(p, n.Rc)
				attachAsLChild(n, g)
				attachAsRChild(n, p)
			} else { // zag-zag
				attachAsRChild(g, p.Lc)
				attachAsRChild(p, n.Lc)
				attachAsLChild(p, g)
				attachAsLChild(n, p)
			}
		}
		if gg != nil {
			if g == gg.Lc {
				attachAsLChild(gg, n)
			} else {
				attachAsRChild(gg, n)
			}
		} else {
			n.Parent = nil
		}
		g.UpdateHeight()
		p.UpdateHeight()
		n.UpdateHeight()
	}
	p = n.Parent
	if p != nil {
		if n.IsLChild() {
			attachAsLChild(p, n.Rc)
			attachAsRChild(n, p)
		} else {
			attachAsRChild(p, n.Lc)
			attachAsLChild(n, p)
		}
		p.UpdateHeight()
		n.UpdateHeight()
	}
	n.Parent = nil
	return n
}

func (s *Splay) Search(e interface{}) *binTree.BinNode {
	n := s.Search_(e)
	if *n == nil {
		n = &s.Hot
	}
	s.Root = s.splay(*n)
	return s.Root
}

func (s *Splay) Insert(e interface{}) *binTree.BinNode {
	if s.Root == nil {
		s.Size_++
		s.Root = binTree.NewNode(e)
	}
	if e == s.Search(e).Data {
		return s.Root
	}
	s.Size_++
	t := s.Root
	if bst.Less(s.Root.Data, e) {
		s.Root = binTree.NewNode(e, nil, t, t.Rc)
		t.Parent = s.Root
		if t.Rc != nil {
			t.Rc.Parent = s.Root
			t.Rc = nil
		}
	} else {
		s.Root = binTree.NewNode(e, nil, t.Lc, t)
		t.Parent = s.Root
		if t.Lc != nil {
			t.Lc.Parent = s.Root
			t.Lc = nil
		}
	}
	t.UpdateHeightAbove()
	return s.Root
}

func (s *Splay) Remove(e interface{}) *binTree.BinNode {
	if s.Root == nil {
		return nil
	}
	if e != s.Search(e).Data {
		return s.Root
	}
	w := s.Root
	if s.Root.Lc == nil {
		s.Root = s.Root.Rc
		if s.Root != nil {
			s.Root.Parent = nil
		}
	} else if s.Root.Rc == nil {
		s.Root = s.Root.Lc
		if s.Root != nil {
			s.Root.Parent = nil
		}
	} else {
		lTree := s.Root.Lc
		lTree.Parent = nil
		s.Root.Lc = nil
		s.Root = s.Root.Rc
		s.Root.Parent = nil
		s.Search(e)
		s.Root.Lc = lTree
		lTree.Parent = s.Root
	}
	s.Size_--
	if s.Root != nil {
		s.Root.UpdateHeight()
	}
	return w
}

func attachAsLChild(p, lc *binTree.BinNode) {
	p.Lc = lc
	if lc != nil {
		lc.Parent = p
	}
}

func attachAsRChild(p, rc *binTree.BinNode) {
	p.Rc = rc
	if rc != nil {
		rc.Parent = p
	}
}

func NewSplay() *Splay {
	return &Splay{
		BST: *bst.NewBST(),
	}
}
